File: day_5/5.js
/**
 * @param {{ type: 'I' | 'R', size: number }[]} shoes
 * @returns {number[]} Available shoes
 */
function organizeShoes(shoes) {
  const result = [];
  const pairShoes = {};
  for (const shoe of shoes) {
    pairShoes[shoe.size] ??= {};
    const otherPair = shoe.type === "I" ? "R" : "I";
    pairShoes[shoe.size][otherPair] ??= 0;
    pairShoes[shoe.size][shoe.type] ??= 0;
    if (pairShoes[shoe.size][otherPair] > 0) {
      pairShoes[shoe.size][otherPair]--;
      result.push(shoe.size);
    } else {
      pairShoes[shoe.size][shoe.type]++;
    }
  }
  return result;
}

console.log(
  organizeShoes([
    { type: "I", size: 38 },
    { type: "R", size: 38 },
    { type: "R", size: 42 },
    { type: "I", size: 41 },
    { type: "I", size: 42 },
  ])
);

console.log(
  organizeShoes([
    { type: "I", size: 38 },
    { type: "R", size: 38 },
    { type: "I", size: 38 },
    { type: "I", size: 38 },
    { type: "R", size: 38 },
  ])
);

File: day_5/5.py
def organizeShoes(shoes):
    array_of_sizes = []
    shoes_dict = {}
    for shoe in shoes:
        size = shoe["size"]
        type = shoe["type"]
        other_pair = "I" if (type == "R") else "R"
        shoes_dict.setdefault(size, {})
        shoes_dict[size].setdefault(type, 0)
        shoes_dict[size].setdefault(other_pair, 0)
        if shoes_dict[size].get(other_pair) > 0:
            array_of_sizes.append(size)
            shoes_dict[size][other_pair] -= 1
        else:
            shoes_dict[size][type] += 1
    return array_of_sizes


print(
    organizeShoes(
        [
            {"type": "I", "size": 38},
            {"type": "R", "size": 38},
            {"type": "R", "size": 42},
            {"type": "I", "size": 41},
            {"type": "I", "size": 42},
        ]
    )
)
